### DISCUSSION TITLE
Xyphro UsbGpib (Old & New Firmware) vs Agilent 82357B vs AR488 Arduino-GPIB Read Performance Comparison

### DISCUSSION THREAD

**Discussion by Niravk1997** [xyphro/UsbGpib#47] [General]
So, I am creating an open-source scope software (not yet released) for the Tektronix TDS 500 600 700 series as the obsolete official scope software (Tektronix WaveStar) does not work with open source GPIB adapters.

My software is designed to capture waveforms as fast as possible. It is heavily multi-threaded. While one thread downloads the waveform from the scope and inserts it into a queue. Another thread removes the waveform from the said queue and processes them. The waveform capture and waveform process both happen independent of each other. The waveform download process does not have to wait for anything.

I stopped the scope from acquiring new waveforms before capturing the waveforms, don't want the scope triggering possibly affecting the benchmark results.

PC Specs: Ryzen 9 7900 @ RTX 3090 @ 32GB DDR5 6000MHz @ Windows 10 Pro

I decided to benchmark my GPIB adapters, below are the results.

![Results](https://github.com/xyphro/UsbGpib/assets/52579388/e88e8f58-d3a8-4651-b34f-fcb0970a62a5)

First, we compare Xyphro UsbGpib old firmware vs new firmware.
![Capture](https://github.com/xyphro/UsbGpib/assets/52579388/dd5ab35e-e7af-4a8f-9932-df83e4f29c15)

The new firmware is much faster compared to the old one, especially when capturing large waveforms.

Then we compare Xyphro UsbGpib (new firmware) vs Agilent 82357B.
![Capture 2](https://github.com/xyphro/UsbGpib/assets/52579388/431527cb-27cd-46e9-89ac-664a910d6ede)

Here we see the Agilent 82357B being a bit faster than Xyphro UsbGpib (new firmware). However, having to deal with Keysight software is a pain.

Lastly, we compare the AR488 Arduino-GPIB with the other. This adapter is by far the cheapest and the easiest to assemble.
![Capture 3](https://github.com/xyphro/UsbGpib/assets/52579388/55ffa51b-d992-4282-8eb9-2523476279cb)

Pictures of the devices tested.
![Devices - Copy](https://github.com/xyphro/UsbGpib/assets/52579388/f5469368-5b3a-4518-b2a8-856946f14c55)

Pictures of the instrument used for capturing waveforms.
![Device](https://github.com/xyphro/UsbGpib/assets/52579388/f2e00300-1e5c-4d9b-9aa8-9e6cde139628)

Picture of NX-Starwave.
![NX-StareWave](https://github.com/xyphro/UsbGpib/assets/52579388/4a3c5d2b-eff1-4458-af62-d7cc346ae234)

In conclusion, I did not encounter any errors or issues when testing the Xyphro UsbGpib (old & new firmware) and the AR488 Arduino-GPIB adapter. The Agilent 82357B however gave a Visa Timeout error when trying to capture an 8Mpts waveform from the scope.

[Results Excel file.xlsx](https://github.com/xyphro/UsbGpib/files/13730015/Results.Excel.file.xlsx)

**Comment by Niravk1997** [comment]
Also, a special thanks to Xyphro for creating an amazing GPIB adpater.

**Reply by xyphro** [reply]
Wow, you impress me a lot: Not even 24h after release and an impressive summary. It is nice to see benchmark data too.
Thanks a lot!

Btw: In case you see issues on any equipment let me know - I tested a lot of old equipment like keithley dmm199 and older already. I want this to work with all equipment. Sometimes it is tricky (like for this dmm199) cause instruments use strange syntaxes for commands or are very critical on termination settings, which is all possible with the adapter but might take some time till it is 100% clear how to control the equipment due to old school documentation. This also applies to other adapters like niusb and is related to the actual instrument.

The next version with ethernet and high speed usb will go on the same level of speed as your agilent adapter, very sure - it is a clear and realistic goal. Already got pcb and components and just wait for some free time to solder and get the first time a debug connection running and start with a fresh plate to structure the code clearer as it is now.

I love the sw that you show, this looks far beyond everything I have seen so far! Will it support tds7104 too or have an abstraction layer to plug other similar tek instruments too?
Also the multithreaded approach is something which is perfect, this will keep the gui very responsive compared to some other solutions coming directly from manufacturers. Who wants to wait x seconds to be able to move a window?!
Looking forward to see it!

**Reply by Niravk1997** [reply]
I usually support instruments that I own, so there won't be any support for tds7104 for this version but maybe the next one. The source code will be available so anyone can chnage it and support any scope they want. but this is for next year. Anyway, can't wait to test the new versions of you adapter. Until next time.

**Comment by fenugrec** [comment]
Cool, thanks for sharing the data @Niravk1997  . Were your test waveforms in Sample mode (1 byte per point) or hi-res / average (2 byte/point) ? Curious to try the same benchmark with my tds744 and stm32-based GPIB adapter (prologix-compatible).

**Reply by Niravk1997** [reply]
Sampled mode (1 byte per point) for the results. For tds744a or lower, the performance won't be good in my opinion.

Also my software downloads waveforms in raw (not ASCII)and process them on the client PC.

**Comment by fenugrec** [comment]
> Sampled mode (1 byte per point) for the results. For tds744a or lower, the performance won't be good in my opinion.

Thanks.
I agree, based on previously testing "simple" operations like GPIB plotting, the 744A is not very fast. Still going to do a quick test just to compare. Will update this post later.


[EDIT] well, speed was quite underwhelming:  ~ 20kB/s. Not sure where the bottleneck is however, could be the 744 or my adapter.

**Reply by xyphro** [reply]
Just FYI: with file transfer SCPI commands and a new adapter I work on I reached 1536KBytes/s for reads without special speed tweaks just yesterday night. Stopped then with a big smile in my face :-)

**Comment by Niravk1997** [comment]
Hey xyphro,

I am including a compiled beta release of my software with this post if you want to test it with your Tektronix tds7104. Your scope is newer than my TDS784D, not sure if the waveform data format is the same or not.

Should run on Windows 10 & 11, must have a Visa software installed. Hopefully I can release the project by end of february.
[NX-StarWave (Beta Release).zip](https://github.com/xyphro/UsbGpib/files/13982569/NX-StarWave.Beta.Release.zip)

VirusTotal on the zip file (https://www.virustotal.com/gui/file/3714e8d462431911dfdc7b22a3b96099ca762956ba65c63bdc249d72bf527251/detection), hmm one false positive.

**Reply by Niravk1997** [reply]
Hmm, shame I just read the programming manual for TDS7104 and Tek unfortunately changed the SCPI commands for getting the waveform. The waveform data format also changed slightly.
